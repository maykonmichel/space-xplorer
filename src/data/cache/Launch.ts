import {FieldPolicy, TypePolicy} from '@apollo/client';
import {format} from 'date-fns';

const launch_date_formatted: FieldPolicy<string> = {
  read: (existing, {readField}) => {
    const date = readField<string>('launch_date_local') || '';
    return format(new Date(date), 'MM/dd/yyyy hh:mmaaa');
  },
};

const Launch: TypePolicy = {
  fields: {
    launch_date_formatted,
  },
};

export default Launch;
