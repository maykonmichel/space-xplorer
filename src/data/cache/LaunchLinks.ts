const LaunchLinks = {
  merge: true,
};

export default LaunchLinks;
