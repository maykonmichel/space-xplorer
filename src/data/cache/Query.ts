import {FieldPolicy, Reference, TypePolicy} from '@apollo/client';
import {sortBy} from 'ramda';

const launchesPast: FieldPolicy<Reference[]> = {
  read: (existing, {readField}) => {
    if (!existing) return;

    return sortBy<Reference>(
      ref => readField<string>('launch_date_local', ref) || '',
      existing,
    );
  },
};

const Query: TypePolicy = {
  fields: {
    launchesPast,
  },
};

export default Query;
