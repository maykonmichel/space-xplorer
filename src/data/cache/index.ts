import {InMemoryCache} from '@apollo/client';
import {format} from 'date-fns';

import {initializeFavoriteImages} from '~/data/favoriteImages';

const cache = new InMemoryCache({
  typePolicies: {
    Launch: {
      fields: {
        launch_date_formatted: {
          read: (existing, {readField}) => {
            const unix = readField<number>('launch_date_unix') || 0;
            return format(unix, 'MM/dd/yyyy hh:mmaaa');
          },
        },
      },
    },
    LaunchLinks: {
      merge: true,
    },
  },
});

export const initializeCache = () => {
  initializeFavoriteImages(cache);
};

initializeCache();

export default cache;
