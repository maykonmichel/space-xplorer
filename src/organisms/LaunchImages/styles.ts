import {StyleSheet} from 'react-native';

export default StyleSheet.create({
  container: {
    paddingVertical: 40,
  },
});
