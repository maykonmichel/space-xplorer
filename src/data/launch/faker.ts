import faker from 'faker';
import {repeat} from 'ramda';

import {Launch} from '~/data/launch';

export const fakerFlickrImages = (): Launch['links']['flickr_images'] =>
  repeat(faker.image.image(), faker.datatype.number({min: 1, max: 5}));

export const fakerLaunch = (): Launch => ({
  id: faker.datatype.number().toString(),
  details: faker.lorem.paragraph(),
  mission_name: faker.lorem.words(),
  launch_date_formatted: faker.date.past().toLocaleDateString(),
  launch_date_local: faker.date.past().toLocaleDateString(),
  links: {
    article_link: faker.internet.url(),
    flickr_images: fakerFlickrImages(),
  },
  rocket: {
    rocket_name: faker.lorem.words(),
  },
});
