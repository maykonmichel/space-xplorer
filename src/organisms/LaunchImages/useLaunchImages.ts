import {useFavoriteImages} from '~/data/favoriteImages';

const MAX = 3;

export default (data: string[]) => {
  const {
    client: {cache},
    data: {favoriteImages = []} = {},
  } = useFavoriteImages();

  const isFavorite = (uri: string) => favoriteImages.includes(uri);

  const toggleFavorite = (uri: string) =>
    cache.modify({
      fields: {
        favoriteImages: () =>
          isFavorite(uri)
            ? favoriteImages.filter(image => image !== uri)
            : [...favoriteImages, uri],
      },
    });

  const slice = data.slice(0, MAX);

  return {
    slice,
    toggleFavorite,
  };
};
