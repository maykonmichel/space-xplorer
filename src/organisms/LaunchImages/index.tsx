import React, {FC} from 'react';
import {FlatList, ListRenderItem} from 'react-native';

import LaunchImage from '~/molecules/LaunchImage';

import useLaunchImages from './useLaunchImages';

const keyExtractor = (uri: string, index: number) => `${uri}_${index}`;

export type Props = {
  data: string[];
};

const LaunchImages: FC<Props> = ({data}) => {
  const {slice, toggleFavorite} = useLaunchImages(data);

  const renderItem: ListRenderItem<string> = ({item}) => (
    <LaunchImage uri={item} onPress={toggleFavorite} />
  );

  return (
    <FlatList
      data={slice}
      keyExtractor={keyExtractor}
      renderItem={renderItem}
      horizontal
    />
  );
};

export default LaunchImages;
