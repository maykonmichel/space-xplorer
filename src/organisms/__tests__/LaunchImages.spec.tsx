import {MockedProvider} from '@apollo/client/testing';
import {render} from '@testing-library/react-native';
import faker from 'faker';
import React from 'react';

import cache, {initializeCache} from '~/data/cache';
import LaunchImages, {Props} from '~/organisms/LaunchImages';

const getSut = (props: Props) =>
  render(
    <MockedProvider cache={cache}>
      <LaunchImages {...props} />
    </MockedProvider>,
  );

describe('<LaunchImages />', () => {
  beforeEach(() => {
    initializeCache();
  });

  it('should render as expected', async () => {
    const data = [faker.image.imageUrl()];
    const {toJSON} = getSut({data});

    expect(toJSON()).toMatchSnapshot();
  });
});
