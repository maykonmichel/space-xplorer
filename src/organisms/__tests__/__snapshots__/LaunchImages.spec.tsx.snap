// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LaunchImages /> should render as expected 1`] = `
<RCTScrollView
  activeAnimationOptions={null}
  activeAnimationType="timing"
  activeSlideAlignment="center"
  activeSlideOffset={20}
  apparitionDelay={0}
  automaticallyAdjustContentInsets={false}
  autoplay={false}
  autoplayDelay={1000}
  autoplayInterval={3000}
  callbackOffsetMargin={5}
  containerCustomStyle={Object {}}
  contentContainerCustomStyle={
    Object {
      "marginVertical": 40,
    }
  }
  contentContainerStyle={
    Array [
      Object {
        "paddingLeft": 75,
        "paddingRight": 75,
      },
      Object {
        "marginVertical": 40,
      },
    ]
  }
  data={
    Array [
      "http://placeimg.com/640/480",
    ]
  }
  decelerationRate="fast"
  directionalLockEnabled={true}
  disableVirtualization={false}
  enableMomentum={false}
  enableSnap={true}
  firstItem={0}
  forwardedRef={[Function]}
  getItem={[Function]}
  getItemCount={[Function]}
  hasParallaxImages={false}
  horizontal={true}
  inactiveSlideOpacity={0.7}
  inactiveSlideScale={0.9}
  inactiveSlideShift={0}
  initialNumToRender={7}
  invertStickyHeaders={false}
  inverted={false}
  itemWidth={600}
  keyExtractor={[Function]}
  layout="default"
  lockScrollTimeoutDuration={1000}
  lockScrollWhileSnapping={false}
  loop={false}
  loopClonesPerSide={3}
  maxToRenderPerBatch={15}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onResponderRelease={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
  overScrollMode="never"
  pinchGestureEnabled={false}
  removeClippedSubviews={true}
  renderItem={[Function]}
  scrollEnabled={true}
  scrollEventThrottle={1}
  scrollsToTop={false}
  shouldOptimizeUpdates={true}
  showsHorizontalScrollIndicator={false}
  showsVerticalScrollIndicator={false}
  slideStyle={Object {}}
  sliderWidth={750}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "flexDirection": "row",
      "opacity": 0,
      "width": 750,
    }
  }
  swipeThreshold={20}
  updateCellsBatchingPeriod={50}
  useScrollView={false}
  vertical={false}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={15}
>
  <View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "opacity": 1,
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
            "width": 600,
          }
        }
      >
        <View
          accessibilityLabel="Toggle favorite"
          accessibilityRole="imagebutton"
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "aspectRatio": 1.7777777777777777,
                  "width": "100%",
                },
              ]
            }
          >
            <FastImageView
              accessibilityRole="image"
              resizeMode="cover"
              source={
                Object {
                  "uri": "http://placeimg.com/640/480",
                }
              }
              style={
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
